File: blob-module/ios/Blob.swift
import Foundation
import ExpoModulesCore

public class Blob: SharedObject {
  var blobParts: [BlobPart]
  var options: BlobOptions
  
  init(blobParts: [BlobPart]?, options: BlobOptions?) {
    self.blobParts = blobParts ?? []
    self.options = options ?? BlobOptions()
  }

  var size: Int {
    return blobParts.reduce(0) { $0 + $1.size() }
  }

  var type: String {
    return options.type
  }
  
  func slice(start: Int = 0, end: Int? = nil, contentType: String? = nil) -> Blob {
      let blobSize = self.size
      var relativeStart = start
    
      if relativeStart < 0 {
          relativeStart = max(blobSize + relativeStart, 0)
      } else {
          relativeStart = min(relativeStart, blobSize)
      }

      var relativeEnd = end ?? blobSize
      if relativeEnd < 0 {
          relativeEnd = max(blobSize + relativeEnd - 1, 0)
      } else {
          relativeEnd = min(relativeEnd, blobSize) - 1
      }

      let span = max(relativeEnd - relativeStart, 0)
      if span == 0 {
          let type = normalizedContentType(contentType)
          return Blob(blobParts: [], options: BlobOptions(type: type, endings: self.options.endings))
      }

      var dataSlice: [BlobPart] = []
      var currentPos = 0
      var remaining = span

      for part in blobParts {
          let partSize = part.size()
          if currentPos + partSize <= relativeStart {
              currentPos += partSize
              continue
          }
          if remaining <= 0 {
              break
          }
          let partStart = max(0, relativeStart - currentPos)
          let partEnd = min(partSize, partStart + remaining)
          let length = partEnd - partStart
          if length <= 0 {
              currentPos += partSize
              continue
          }

          switch part {
          case .string(let str):
              let utf8 = Array(str.utf8)
            let subUtf8 = Array(utf8[partStart...partStart+length])
              if let subStr = String(bytes: subUtf8, encoding: .utf8) {
                  dataSlice.append(.string(subStr))
              }
          case .data(let data):
              let subData = data.subdata(in: partStart..<partStart+length+1)
              dataSlice.append(.data(subData))
          case .typedArray(let typedArray):
              if partStart == 0 && length == typedArray.byteLength {
                  dataSlice.append(.typedArray(typedArray))
              } else {
                  let rawPtr = typedArray.rawPointer.advanced(by: partStart)
                  let buffer = UnsafeBufferPointer<UInt8>(start: rawPtr.assumingMemoryBound(to: UInt8.self), count: length)
                  let newData = Data(buffer: buffer)
                  dataSlice.append(.data(newData))
              }
          case .blob(let blob):
              let subBlob = blob.slice(start: partStart, end: partStart + length)
              dataSlice.append(.blob(subBlob))
          }

          currentPos += partSize
          remaining -= length
      }
    
      let type = normalizedContentType(contentType)

      return Blob(blobParts: dataSlice, options: BlobOptions(type: type, endings: self.options.endings))
  }
  
  func text() -> String {
    return blobParts.reduce("") { $0 + $1.text() }
  }
}

enum EndingType: String, Enumerable {
  case transparent = "transparent"
  case native = "native"
}

struct BlobOptions: Record {
  @Field
  var type: String = ""
  @Field
  var endings: EndingType = .transparent
}

private func normalizedContentType(_ type: String?) -> String {
    guard let type = type, !type.isEmpty else { return "" }
    let ascii = type.utf8.allSatisfy { $0 >= 0x20 && $0 <= 0x7E }
    let invalid = type.contains { $0 == "\u{00}" || $0 == "\u{0A}" || $0 == "\u{0D}" }
    if !ascii || invalid {
        return ""
    }
    return type.lowercased()
}

enum BlobPart {
    case string(String)
    case blob(Blob)
    case data(Data)
    case typedArray(TypedArray)
}


extension BlobPart {
    func text() -> String {
        switch self {
        case .string(let str):
            return str
        case .data(let data):
            return String(decoding: data, as: UTF8.self)
        case .blob(let blob):
            return blob.text()
        case .typedArray(let typedArray):
          let buffer = UnsafeBufferPointer<UInt8>(
               start: typedArray.rawPointer.assumingMemoryBound(to: UInt8.self),
               count: typedArray.byteLength
           )
          return String(decoding: buffer, as: UTF8.self)
        }
    }
    
    func size() -> Int {
        switch self {
        case .string(let str):
            return str.lengthOfBytes(using: .utf8)
        case .data(let data):
             return data.count
        case .blob(let blob):
            return blob.size
        case .typedArray(let typedArray):
            return typedArray.byteLength
        }
    }
}

File: blob-module/android/src/main/java/expo/modules/blobmodule/Blob.kt
package expo.modules.blobmodule

import android.util.Log
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import expo.modules.kotlin.sharedobjects.SharedObject
import expo.modules.kotlin.typedarray.TypedArray
import expo.modules.kotlin.types.EitherOfThree
import expo.modules.kotlin.types.Enumerable

class Blob() : SharedObject() {
    var blobParts: List<InternalBlobPart> = listOf()
    var size: Int = 0
    var options: BlobOptions = BlobOptions()

    constructor(blobParts: List<InternalBlobPart>, options: BlobOptions = BlobOptions()) : this() {
        this.blobParts = blobParts
        this.options = options

        for (bp in blobParts) {
            size += bp.size()
        }
    }

    fun text(): String {
        var str = ""
        for (bp in blobParts) {
            str += bp.text()
        }
        return str
    }

    private fun InternalBlobPart.offsetSlice(start: Int, end: Int, offset: Int): InternalBlobPart {
        var s: Int = start - offset
        var e: Int = end - offset
        if (s < 0) {
            s = 0
        }
        if (e > this.size()) {
            e = this.size()
        }
        return when (this) {
            is InternalBlobPart.StringPart -> InternalBlobPart.StringPart(string.substring(s, e))
            is InternalBlobPart.BlobPart -> InternalBlobPart.BlobPart(blob.slice(s, e, ""))
            is InternalBlobPart.BufferPart -> InternalBlobPart.BufferPart(
                buffer.slice(s..<e).toByteArray()
            )
        }
    }

    fun slice(start: Int, end: Int, contentType: String): Blob {
        var i: Int = 0
        var bps: MutableList<InternalBlobPart> = mutableListOf()

        for (bp in blobParts) {
            if (i + bp.size() <= start) {
                i += bp.size()
                continue
            }
            if (i >= end) {
                break
            }
            bps.add(bp.offsetSlice(start, end, i))
            i += bp.size()
        }

        return Blob(bps, BlobOptions(contentType))
    }
}

typealias BlobPart = EitherOfThree<String, Blob, TypedArray>

private fun TypedArray.bytes(): ByteArray {
    var ba = ByteArray(this.byteLength)

    for (i in 0..<this.byteLength) {
        ba[i] = this.readByte(i)
    }

    return ba
}

const val CR = '\r'
const val LF = '\n'
fun String.toNativeNewlines(): String {
    var i = 0
    var str = ""

    while (i < this.length) {
        if (this[i] == CR) {
            str += LF
            i += 1
            if (i < this.length && this[i] == LF) {
                i += 1
            }
            continue
        }
        str += this[i]
        i += 1
    }
    return str
}

fun List<BlobPart>.internal(nativeNewlines: Boolean): List<InternalBlobPart> {
    return this.map() { bp: BlobPart ->
        if (bp.`is`(String::class)) {
            bp.get(String::class).let {
                val str = if (nativeNewlines) {
                    it.toNativeNewlines()
                } else {
                    it
                }
                InternalBlobPart.StringPart(str)
            }
        } else if (bp.`is`(Blob::class)) {
            bp.get(Blob::class).let {
                InternalBlobPart.BlobPart(it)
            }
        } else {
            bp.get(TypedArray::class).let {
                InternalBlobPart.BufferPart(it.bytes())
            }
        }
    }
}

sealed class InternalBlobPart() {
    data class StringPart(val string: String) : InternalBlobPart()
    data class BlobPart(val blob: Blob) : InternalBlobPart()
    data class BufferPart(val buffer: ByteArray) : InternalBlobPart()

    fun size(): Int {
        return when (this) {
            is StringPart -> string.length
            is BlobPart -> blob.size
            is BufferPart -> buffer.size
        }
    }

    fun text(): kotlin.String {
        return when (this) {
            is StringPart -> string
            is BlobPart -> blob.text()
            is BufferPart -> buffer.decodeToString()
        }
    }
}

enum class EndingType(val str: String = "transparent") : Enumerable {
    TRANSPARENT("transparent"), NATIVE("native"),
}

data class BlobOptions(val type: String = "", val endings: EndingType = EndingType.TRANSPARENT)

class BlobOptionsBag : Record {
    @Field
    val type: String = ""

    @Field
    val endings: EndingType = EndingType.TRANSPARENT
}
